src/main/java/gregtech/api/capability/impl/AbstractRecipeLogic.java

```diff
@@ -368,27 +369,45 @@ public void forceRecipeRecheck() {
      */
     protected void trySearchNewRecipe() {
         long maxVoltage = getMaxVoltage();
-        Recipe currentRecipe;
+        Recipe currentRecipe = null;
+        Iterator<Recipe> recipeIterator;
         IItemHandlerModifiable importInventory = getInputInventory();
         IMultipleTankHandler importFluids = getInputTank();
 
         // see if the last recipe we used still works
         if (checkPreviousRecipe()) {
             currentRecipe = this.previousRecipe;
             // If there is no active recipe, then we need to find one.
-        } else {
-            currentRecipe = findRecipe(maxVoltage, importInventory, importFluids);
         }
-        // If a recipe was found, then inputs were valid. Cache found recipe.
-        if (currentRecipe != null) {
+        recipeIterator = getRecipeIterator(maxVoltage, importInventory, importFluids);
+
+        // proceed if previous recipe still works.
+        if (currentRecipe != null && checkRecipe(currentRecipe) && prepareRecipe(currentRecipe)) {
             this.previousRecipe = currentRecipe;
+            return;
         }
-        this.invalidInputsForRecipes = (currentRecipe == null);
 
-        // proceed if we have a usable recipe.
-        if (currentRecipe != null && checkRecipe(currentRecipe)) {
-            prepareRecipe(currentRecipe);
+        // Search for a new recipe if the previous one is no longer valid
+        while (recipeIterator != null && recipeIterator.hasNext()) {
+            Recipe next = recipeIterator.next();
+            if (next == null) continue;
+
+            // since previous attempts to prepare a recipe will have flagged either the input
+            // or outputs as invalid we must reset the flags before attempting to prepare another recipe
+            this.isOutputsFull = false;
+            this.invalidInputsForRecipes = false;
+
+            if (checkRecipe(next) && prepareRecipe(next)) {
+                // If a new recipe was found, cache found recipe.
+                this.previousRecipe = next;
+                return;
+            }
         }
+
+        // if no valid recipes are found mark the inputs and outputs as invalid so any changes
+        // will re-trigger a recipe search
+        this.isOutputsFull = true;
+        this.invalidInputsForRecipes = true;
```

Comment:
Is there a better way to know which of these needs to be marked? Marking both removes some performance benefits of having them separated

Reply:
fixed